File: topological-sort/topological_sort.rb
def topological_sort(graph)
  visited = Set.new
  stack = []
  
  def dfs(node)
    return if visited.include?(node)
    visited.add(node)
    (graph[node] || []).each { |neighbor| dfs(neighbor) }
    stack.push(node)
  end
  
  graph.keys.each { |node| dfs(node) }
  stack.reverse
end

puts topological_sort({0 => [1, 2], 1 => [3], 2 => [3], 3 => []}).inspect

File: topological-sort/topological_sort.rs
use std::collections::{HashMap, HashSet};

fn topological_sort(graph: &HashMap<i32, Vec<i32>>) -> Vec<i32> {
    let mut visited = HashSet::new();
    let mut stack = Vec::new();
    
    fn dfs(node: i32, graph: &HashMap<i32, Vec<i32>>, visited: &mut HashSet<i32>, stack: &mut Vec<i32>) {
        if visited.contains(&node) {
            return;
        }
        
        visited.insert(node);
        
        if let Some(neighbors) = graph.get(&node) {
            for &neighbor in neighbors {
                dfs(neighbor, graph, visited, stack);
            }
        }
        
        stack.push(node);
    }
    
    for &node in graph.keys() {
        dfs(node, graph, &mut visited, &mut stack);
    }
    
    stack.reverse();
    stack
}

fn main() {
    let mut graph = HashMap::new();
    graph.insert(0, vec![1, 2]);
    graph.insert(1, vec![3]);
    graph.insert(2, vec![3]);
    graph.insert(3, vec![]);
    
    println!("{:?}", topological_sort(&graph));
}